libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/ParentChildrenSyncDownTarget.kt

```diff
@@ -0,0 +1,363 @@
+            //
+            // This is the soql to identify children
+
+            // We are doing
+            //  select Id, (select Id from children) from Parents where soqlParentFilter
+            // It could be better to do
+            //  select Id from child where qualified-soqlParentFilter (e.g. if filter is Name = 'A' then we would use Parent.Name = 'A')
+            // But "qualifying" parentSoqlFilter without parsing it could prove tricky
+
+            // Nested query
+            val nestedFields: MutableList<String> = ArrayList()
+            nestedFields.add(childrenInfo.idFieldName)
+            val builderNested: SOQLBuilder =
+                SOQLBuilder.getInstanceWithFields(nestedFields)
+            builderNested.from(childrenInfo.sobjectTypePlural)
+
+            // Parent query
+            val fields: MutableList<String> = ArrayList()
+            fields.add(idFieldName)
+            fields.add("(${builderNested.build()})")
+            val builder: SOQLBuilder = SOQLBuilder.getInstanceWithFields(fields)
+            builder.from(parentInfo.sobjectType)
+            builder.where(parentSoqlFilter)
+            return builder.build()
+        }
+
+    @Throws(JSONException::class, IOException::class)
+    override fun cleanGhosts(syncManager: SyncManager, soupName: String, syncId: Long): Int {
+        // Taking care of ghost parents
+        val localIdsSize = super.cleanGhosts(syncManager, soupName, syncId)
+
+        // Taking care of ghost children
+
+        // NB: ParentChildrenSyncDownTarget's getNonDirtyRecordIdsSql does a join between parent and children soups
+        // We only want to look at the children soup, so using SoqlSyncDownTarget's getNonDirtyRecordIdsSql
+        val localChildrenIds: MutableSet<String> = getIdsWithQuery(
+            syncManager, super.getNonDirtyRecordIdsSql(
+                childrenInfo.soupName,
+                childrenInfo.idFieldName,
+                buildSyncIdPredicateIfIndexed(syncManager, childrenInfo.soupName, syncId)
+            )
+        )
+        val remoteChildrenIds = getChildrenRemoteIdsWithSoql(syncManager, soqlForRemoteChildrenIds)
+        localChildrenIds.removeAll(remoteChildrenIds)
+        if (localChildrenIds.size > 0) {
+            deleteRecordsFromLocalStore(
+                syncManager,
+                childrenInfo.soupName,
+                localChildrenIds,
+                childrenInfo.idFieldName
+            )
+        }
+        return localIdsSize
+    }
+
+    @Throws(IOException::class, JSONException::class)
+    protected fun getChildrenRemoteIdsWithSoql(
+        syncManager: SyncManager,
+        soqlForChildrenRemoteIds: String
+    ): Set<String> {
+
+        // Makes network request and parses the response.
+        var records = startFetch(syncManager, soqlForChildrenRemoteIds)
+        val remoteChildrenIds: MutableSet<String> = HashSet(parseChildrenIdsFromResponse(records))
+        while (true) {
```

Comment:
@wmathurin - Is this intended to be `while (true)`?

Reply:
There is a break line 193 when records is null. I'll rewrite in a more readable way.

Reply:
Addressed with https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2459/commits/4a71a36046f63c109ee9f836ec01999a4f543c86